Add documentation: How to build a lexical model (#51)
keymanapp/lexical-models: docs/how-to-build-a-lexical-model.md

---
Author: Eddie Antonio Santos
Date: July 22, 2019
Keyman Version: 12.0.7-alpha-e58173f
---

How to build a lexical model
============================

A **lexical model** or a **dictionary** is a way to enable **predictive text**
for your keyboard.

Requirements
------------

You will need to have:

 - [Node.js][] >= 10.0
 - (Windows only) [Git Bash](https://gitforwindows.org/)

> [Node.js]: https://nodejs.org/en/

Reviewer: Might as well put that link inline if you doing that for Git Bash anyway (but formatting will change when it moves...)

create this TSV file.

[Google Sheets]: https://sheets.google.com/
[Microsoft Excel]: https://products.office.com/en/excel

### Example

I have words in my language of choice, SENĆOŦEN.

Here is my list of words, with the count of how many times I've seen the
word:

| Word  | Count   |
|-------|---------|
| TŦE   | 13644   |
| E     | 9134    |
| SEN   | 4816    |
| Ȼ     | 3479    |
| SW̱    | 2621    |
| NIȽ   | 2314    |
| U¸    | 2298    |
| I¸    | 1988    |
| ȻSE   | 1925    |
| I     | 1884    |

I've entered this information into my spreadsheet of choice, [Google
Sheets][]. I've shared this spreadsheet publicly [here][sencoten-sheet].
The order of the columns matters:

The first column **MUST** be the words, and the second column **MUST**
be the counts. Additional columns are ignored. The spreadsheet can also
be a single column of all of the words in the language.

This is what my word list looks like in Google Sheets:

![The word list, as it appears in Google Sheets](./sencoten-sheets-full.png).

[sencoten-sheet]: https://docs.google.com/spreadsheets/d/10zhIc439BCSSooL_-HeJ6TUHd-ovkiXYcIGe-pHDTSg/edit?usp=sharing

> zsh:1: no matches found: https://docs.google.com/spreadsheets/d/10zhIc439BCSSooL_-HeJ6TUHd-ovkiXYcIGe-pHDTSg/edit?usp=sharing

Reviewer: Is this supposed to be here? See also next line!

Author: haha, I think I keysmashed in Vim here...

zsh:1: no matches found: https://docs.google.com/spreadsheets/d/10zhIc439BCSSooL_-HeJ6TUHd-ovkiXYcIGe-pHDTSg/edit?usp=sharing

Now, we download the spreadsheet in the [required
format](#appendix:tsv). To do this, in Google Sheets, select "File"
» "Download as" » "Tab-separated values (.tsv, current sheet)".

![Exporting the TSV file from Google Sheets](./sencoten-sheets-save-as.png)

I'll save mine as **wordlist.tsv**.

Step 2: Create a unique model ID
--------------------------------

Now that we have our TSV file, called **wordlist.tsv**, let's create the
associated packaging folder and data.

Our model needs a **unique ID**. The lexical model ID is a string in the
format:

> _author_._bcp47_._uniq_

### Author

_author_ is an identifier for the maintainer of the model. This
identifier is a string of one or more ASCII lowercase letters, digits, and the
underscore (`_`), and it must begin with a lowercase ASCII letter. Some

> organizations have a author identifier already, like `sil` for SIL

Reviewer: ```suggestion
organizations have an author identifier already, like `sil` for SIL
```

replace them with underscores; likewise, replace uppercase letters with
lowercase letters.

Typically, this will just be the [ISO 639-3][] code of your language,
and nothing else. In the case of SENĆOŦEN, [its ISO 639-3 code is
`str`](https://iso639-3.sil.org/code/str).

In some cases, you will have to specify the script (writing system), and
the region. For example, if I'm creating a model for Plains Cree (ISO
639-3: `crk`) it can be written both in the Latin script ([ISO 15924][]:
`Latn`) and in Unified Canadian Aboriginal Syllabics ([ISO 15924][]:
`Cans`). Say my model is only designed for syllabics. The correct BCP 47 tag
is:

    crk-Cans

However, to create a valid lexical model ID, we need to replace the
dash with an underscores, and convert all uppercase letters to
lowercase. Thus, our `bcp47` is:

    crk_cans

[ISO 639-3]: https://iso639-3.sil.org/code_tables/639/data
[ISO 15924]: https://en.wikipedia.org/wiki/ISO_15924
[BCP 47]: https://tools.ietf.org/html/bcp47

### Uniq

_uniq_ is an arbitrary tag that uniquely identifies your specific model.
This identifier can be anything, as long as it hasn't already been used

> before for this specific language. As with `author`

Reviewer: Uniq only needs to be unique to the author, for that language; it doesn't technically need to be unique to other people.

Also note that this file naming scheme is only mandated for models in the shared repo. For private distribution you should be able to call it `foo` (even if that isn't recommended!)

Reviewer 2: I'd find a more concrete example much more helpful. For example, Gene Rodgers of the National Research Council Canada is developing a lexical model for Straits Salish.
- author is `nrc` (for the National Research Council Canada)
- the BCP 47 code for the language is `str`
- the unique code is `grodgers` (or some other convention established by the NRC for uniquely identifying authors within the NRC)
- result is nrc.str.grdogers

Joe Bloggs is developing a lexical model for Hausa written with Arabic script.
- author is `bloggs`
- the BCP 47 code for the language is `ha_arab`
- the unique code is `joe` (or some other string that uniquely identifies Joe from other authors named `Bloggs`)
- result is bloggs.ha_arab.joe

Reviewer: That's not quite right :)

The author is already unique -- if two developers named Bloggs turned up wanting to use the `bloggs` subtag, we'd ask the second one to pick a different author subtag, (so the second developer may get `alfredbloggs` for the author subtag. The `uniq` subtag is really about disambiguating multiple models for the same language from the same author, so for example:

Joe Bloggs has a Straits Salish model based on the Montler wordlist, and another one based on the Salish School Board wordlist (this is semi-fictional but ...). He wants both of them available because of different preferences in the community. So he picks a uniq subtag for each accordingly, for example:

* `bloggs.str.montler`
* `bloggs.str.schoolboard`

In practice, I hope there won't be many models for the same language. Note that dialect differences should really be picked up in the BCP 47 subtag (e.g. en_us, en_ca, en_au)!

Reviewer 2: I appreciate the clarification. I'm still trying to wrap my head around this, and a concrete example helps -- especially when it shows that my understanding was wrong!!

Reviewer: No problemo. The intent is to try to avoid running into namespacing issues which many projects (including keyman keyboards) have run into in the past. This may be a little over the top in some respects but isn't too hard, hopefully, for people to get their head around. The `uniq` subtag could also just be described as a name for the model.

identifier is one or more ASCII lowercase letters, digits, and the underscore
(`_`), and it must start with a lowercase ASCII letter.

### Example

My author is called `example`, and the BCP 47 for my language is `str`.
I'll give my model the unique tag of `tutorial`. Thus my unique model
is:

    example.str.tutorial

Step 4: Creating the package files
----------------------------------

Now that we have a unique ID for our model, and wordlist data, we can
the appropriate packaging files.

Say we're currently in a folder called `release`: our model will follow
this template for its directory structure.

```
release/
└── {author}/
    └── {bcp47}.{uniq}/
        ├── {author}.{bcp47}.{uniq}.model_info
        └── source/
            ├── {author}.{bcp47}.{uniq}.model.kps
            ├── wordlist.tsv
            └── model.ts
```

Notice how under `release`, there is a folder that is named after the
author. Within that folder is a folder for each models created by that
particular author. Within each model folder, this is a folder called
`source` that contains **wordlist.tsv**. We will discuss what the other
files are called a bit later in this document.

> 

Reviewer 2: change "each models" to "each model"

"this is a folder..." Did you mean "there is a folder..."?

For our example of `example.str.tutorial`, I'll want the following
folder structure:

```
release/
└── example/
    └── str.tutorial/
        ├── example.str.tutorial.model_info
        └── source/
            ├── example.str.tutorial.model.kps
            ├── wordlist.tsv
            └── model.ts
```

To create this folder structure, I can execute the following in
Git Bash or the terminal:

    cd /type/the/path/to/release
    mkdir -p example/str.tutorial/source

Now move the wordlist to the appropriate place:

    mv /type/the/current/path/to/wordlist.tsv example/str.tutorial/source/

I now have the following folder structure:

```
.
└── example
    └── str.tutorial
        └── source
            └── wordlist.tsv
```

We need to create the remaining files:
 - The "model info" file: `example.str.tutorial.model_info`
 - The package file: `example.str.tutorial.model.kps`
 - The model configuration file: `model.ts`

> Step 4.1: Creating the model info file

Reviewer: model_info should only be required when building within the repo and distributing via keyman cloud (not used for private, ad-hoc distribution; worth noting this). We'll be adding outside-repo builds before release, I hope!

--------------------------------------

The model info contains a little bit of metadata that is required when
distributing packages. It is a JSON file that lists the model license
and a short description of the lexical model.

The `.model_info` file should follow this template:

```json
{
    "license": "<SPDX license identifier>",
    "description": "<short description of the lexical model>"
}
```

Copy this template, and use it to create your own `.model_info` file.

Save this file to `{author}/{bcp47}.{uniq}/{author}.{bcp47}.{uniq}.model_info`.

> NOTE: the only current valid value for `license` is `"mit"`.

In our `example.str.tutorial` example, I'll use the following
`.model_info`.

```json
{
    "license": "mit",
    "description": "Predicts words in the SENĆOŦEN language (Saanich Dialect)"
}
```

I'll save this file to `example/str.tutorial/example.str.tutorial.model_info`.

Step 4.2: Creating the KPS file
-------------------------------

> **TODO:**

Reviewer: We'll generally encourage model developers to use Keyman Developer to create this file. It saves on silly errors. Clean support for models in Developer however is of course TBD.

  </System>
  <Options>
    <FollowKeyboardVersion/>
  </Options>
  <Info>
    <Name URL="">SENĆOŦEN (Saanich Dialect) Lexical Model</Name>
    <Copyright URL="">© 2019 National Research Council Canada</Copyright>
    <Author URL="mailto:Eddie.Santos@nrc-cnrc.gc.ca">Eddie Antonio Santos</Author>
    <Version>1.0.3</Version>
  </Info>
  <Files>
    <File>
      <Name>..\build\nrc.str.sencoten.model.js</Name>
      <Description>Lexical model nrc.str.sencoten.model.js</Description>
      <CopyLocation>0</CopyLocation>
      <FileType>.model.js</FileType>
    </File>
  </Files>
  <LexicalModels>
    <LexicalModel>
      <Name>SENĆOŦEN dictionary</Name>
      <ID>nrc.str.sencoten</ID>
      <Version>1.0.3</Version>
      <Languages>
        <Language ID="str">North Straits Salish</Language>
        <Language ID="str-Latn">SENĆOŦEN</Language>
      </Languages>
    </LexicalModel>
  </LexicalModels>
</Package>
```

Step 4.3: Creating the model file
---------------------------------

Finally, the `model.ts` is a TypeScript source code file that is
executed to build the model. For our purposes, we can simply copy-paste
the following example:

```typescript

> import LexicalModelCompiler from "@keymanapp/developer-lexical-model-compiler";

Reviewer: Also note: when we integrate as part of Keyman Developer, we'll be making this compiler available as part of the install process.

Author: Could you please elaborate on this point?

Reviewer: Once we have Keyman Developer capable of doing lexical model builds, there won't be a need to hand-write all parts of these files. It will include a template for a model file to start people off and will make sure the compiler is readily available without needing to clone the repo.

I guess the comment was more a reflection on the current reality that that import statement requires a clone of the keymanapp/keyman repo which is somewhat less than ideal!

```

Most importantly, this file specifies that model will use the word list
in `wordlist.tsv`.

For simple lexical models with exactly one wordlist called
`wordlist.tsv`, **this file does not need to be changed any further**.

However, it specifies a few other things:

  - `format: 'trie-1.0'`: Use the trie back-end, that allows for
    prediction based on prefixes of words.
  - `wordBreaking: 'default'`: use Unicode's default word boundary rules
    to find where "words" are when predicting text. This works for most
    languages and writing systems, however, it does not work for
    languages/writing systems that do not mark breaks between words,
    e.g., Chinese, Thai, Laos, Khmer, and Japanese, to name a few. For
    these writing systems, you must provide a language-specific word
    breaking algorithm to find the words in running text.

Step 5: Build the model
-----------------------

Appendix: Tab-separated wordlist file format specification
----------------------------------------------------------

<a id="appendix:tsv"></a>

[appendix]: ./#appendix:tsv

The **lexical model compiler** expects wordlists to abide by the
following **tab-separated values** (TSV) format:

>  - the file is must be **UTF-8** encoded text file

Reviewer: ```suggestion
 - the file must be **UTF-8** encoded text file without Byte Order Mark (BOM).
```

Note: Notepad until recently always prepended a BOM. This is now changing: https://www.bleepingcomputer.com/news/microsoft/windows-10-notepad-is-getting-better-utf-8-encoding-support/

Reviewer 3: I mean, is it hard to strip out the BOM?  Seems like a big ask for those not so familiar with byte encoding particularities.

Author: Good point! It's not really that difficult, and I don't _think_ there's any problem here. I don't think... please debate it over here: keymanapp/keyman#1916!

 - newlines are either **LF** or **CRLF**
 - the file either consists of a **comment** or an **entry**
 - **comment** lines MUST start with the '#' character on the very first column
 - **entries** are one or two columns, separated by the (horizontal)
   **tab character** (U+0009)

>  - column 1 (REQUIRED): the **word form**: can have any character

Reviewer: Suggest noting that quote characters are not used to surround the text?

Author: Correct!

   except tab, CR, or LF. Surrounding whitespace characters are trimmed.
 - column 2 (optional): the **count**: a non-negative integer specifying how many
   times this entry has appeared in the corpus. Blank means
   'indeterminate', and is treated as if the word exists in the corpus,
   but will be predicted at the lowest possible priority.

>  - the following columns are ignored.

Reviewer: ```suggestion
 - any other columns are ignored.
```

> Source:
> [build-trie.js@e58173f](https://github.com/keymanapp/keyman/blob/307436e7c24caa5c720b272640184362c4dc3223/developer/js/lexical-model-compiler/build-trie.ts#L21-L66)

Exporting a spreadsheet from Google Sheets as a TSV will produce
properly formatted output. I have not found a reliable output option in

> Microsoft Excel for Mac that will produce output in the expected format.

Reviewer: Excel for Windows can load the .tsv file okay., but Save has trouble with encoding. However, Copy/Paste into Notepad++ works (older versions of Notepad would work except they prepend a UTF-8 BOM 😭; new versions don't do this).
